File: user-service/src/main/java/socialapp/userservice/service/impl/RelationServiceImpl.java
package socialapp.userservice.service.impl;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import socialapp.userservice.common.exception.InvalidSubscription;
import socialapp.userservice.model.dto.RelationDto;
import socialapp.userservice.model.dto.UserRelationsDto;
import socialapp.userservice.model.entity.Block;
import socialapp.userservice.model.entity.Friendship;
import socialapp.userservice.model.entity.Subscription;
import socialapp.userservice.repository.BlockRepository;
import socialapp.userservice.repository.FriendshipRepository;
import socialapp.userservice.repository.SubscriptionRepository;
import socialapp.userservice.service.RelationService;
import socialapp.userservice.service.UserService;

import java.util.stream.Collectors;

@Service
@Slf4j
@RequiredArgsConstructor
public class RelationServiceImpl implements RelationService {

    private final SubscriptionRepository subscriptionRepository;
    private final BlockRepository blockRepository;
    private final FriendshipRepository friendshipRepository;
    private final UserService userService;

    @Override
    public void deleteBlock(Long id) {
        blockRepository.deleteById(id);
    }

    @Override
    public void deleteSubscription(Long id) {
        subscriptionRepository.deleteById(id);
    }

    @Override
    public void deleteFriendship(Long id, Long by) { // todo refactor govno_code
        var friendship = friendshipRepository.findByID(id);

        var unsubscribedUser = friendship.getSender().getId().equals(by) ?
                friendship.getSender() : friendship.getReceiver();

        if (!unsubscribedUser.getId().equals(by)) {
            return;//throw exception and handle with rest controller advice
        }

        friendshipRepository.deleteById(id);

        var subscription = new Subscription();
        subscription.setReceiver(unsubscribedUser);
        subscription.setSender(friendship.getSender().getId().equals(unsubscribedUser.getId()) ? friendship.getReceiver() : friendship.getSender());

        subscriptionRepository.save(subscription);
    }

    @Override
    public void blockUser(Long sender, Long receiver) {
        var senderUser = userService.findById(sender);
        var receiverUser = userService.findById(receiver);

        var block = new Block();
        block.setSender(senderUser);
        block.setReceiver(receiverUser);

        blockRepository.save(block);
    }

    @Override
    public void subscribeUser(Long sender, Long receiver) {
        var senderUser = userService.findById(sender);
        var receiverUser = userService.findById(receiver);

        var subscription = new Subscription();
        subscription.setSender(senderUser);
        subscription.setReceiver(receiverUser);

        subscriptionRepository.save(subscription);
    }

    @Override
    public void addFriendship(Long sender, Long receiver) {
        var subscription = subscriptionRepository.findBySenderIdAndReceiverId(sender, receiver)
                .orElseThrow(InvalidSubscription::new);

        var friendShip = new Friendship();
        friendShip.setSender(subscription.getSender());
        friendShip.setReceiver(subscription.getReceiver());

        friendshipRepository.save(friendShip);
    }

    @Override
    public UserRelationsDto findUserRelations(Long userId) {
        var user = userService.findById(userId);
        var blockSet = blockRepository.findAllBySenderId(user.getId());
        var incomingSubscriptions = subscriptionRepository.findAllByReceiverId(user.getId());
        var outgoingSubscriptions = subscriptionRepository.findAllBySenderId(user.getId());
        var friends = friendshipRepository.findAllByReceiverIdOrSenderId(user.getId(), user.getId());

        return new UserRelationsDto( // never do this // todo refactor code
                blockSet.stream().map(b -> new RelationDto(b.getReceiver(),b.getCreatedTime().toLocalDate())).collect(Collectors.toSet()),
                friends.stream().map(f -> new RelationDto(f.getReceiver().getId().equals(user.getId()) ? f.getSender() : f.getReceiver(),f.getCreatedTime().toLocalDate()))
                        .collect(Collectors.toSet()),
                outgoingSubscriptions.stream().map(b -> new RelationDto(b.getReceiver(),b.getCreatedTime().toLocalDate())).collect(Collectors.toSet()),
                incomingSubscriptions.stream().map(b -> new RelationDto(b.getSender(),b.getCreatedTime().toLocalDate())).collect(Collectors.toSet())
        );
    }
}

File: user-service/src/main/kotlin/socialapp/ktuserservice/service/impl/RelationServiceImpl.kt
package socialapp.ktuserservice.service.impl

import org.springframework.stereotype.Service
import socialapp.ktuserservice.common.exception.InvalidSubscription
import socialapp.ktuserservice.model.dto.RelationDto
import socialapp.ktuserservice.model.dto.UserRelationsDto
import socialapp.ktuserservice.model.entity.Block
import socialapp.ktuserservice.model.entity.Friendship
import socialapp.ktuserservice.model.entity.Subscription
import socialapp.ktuserservice.repository.BlockRepository
import socialapp.ktuserservice.repository.FriendshipRepository
import socialapp.ktuserservice.repository.SubscriptionRepository
import socialapp.ktuserservice.service.RelationService
import socialapp.ktuserservice.service.UserService
import java.util.stream.Collectors

@Service
class RelationServiceImpl(
    private val subscriptionRepository: SubscriptionRepository,
    private val blockRepository: BlockRepository,
    private val friendshipRepository: FriendshipRepository,
    private val userService: UserService
) : RelationService {

    override fun deleteBlock(id: Long) {
        blockRepository.deleteById(id)
    }

    override fun deleteSubscription(id: Long) {
        subscriptionRepository.deleteById(id)
    }

    override fun deleteFriendship(id: Long, by: Long) {
        val friendship = friendshipRepository.findByID(id)

        val unsubscribedUser = if (friendship.sender?.id == by) friendship.sender else friendship.receiver

        if (unsubscribedUser?.id != by) {
            return
        }

        friendshipRepository.deleteById(id)

        val subscription = Subscription().apply {
            receiver = unsubscribedUser
            sender = if (friendship.sender?.id == unsubscribedUser.id) friendship.receiver else friendship.sender
        }

        subscriptionRepository.save(subscription)
    }

    override fun blockUser(sender: Long, receiver: Long) {
        val senderUser = userService.findById(sender)
        val receiverUser = userService.findById(receiver)

        val block = Block().apply {
            this.sender = senderUser
            this.receiver = receiverUser
        }

        blockRepository.save(block)
    }

    override fun subscribeUser(sender: Long, receiver: Long) {
        val senderUser = userService.findById(sender)
        val receiverUser = userService.findById(receiver)

        val subscription = Subscription().apply {
            this.sender = senderUser
            this.receiver = receiverUser
        }

        subscriptionRepository.save(subscription)
    }

    override fun addFriendship(sender: Long, receiver: Long) {
        val subscription = subscriptionRepository.findBySenderIdAndReceiverId(sender, receiver)
            ?: throw InvalidSubscription()

        val friendShip = Friendship().apply {
            this.sender = subscription.sender
            this.receiver = subscription.receiver
        }


        friendshipRepository.save(friendShip).let {
            subscriptionRepository.delete(subscription)
        }
    }

    override fun findUserRelations(userId: Long): UserRelationsDto {
        val user = userService.findById(userId)
        val blockSet = blockRepository.findAllBySenderId(user.id!!)
        val incomingSubscriptions = subscriptionRepository.findAllByReceiverId(user.id!!)
        val outgoingSubscriptions = subscriptionRepository.findAllBySenderId(user.id!!)
        val friends = friendshipRepository.findAllByReceiverIdOrSenderId(user.id!!, user.id!!)

        return UserRelationsDto(
            blockSet.stream().map { b -> RelationDto(b.id!!, b.receiver!!, b.createdTime?.toLocalDate()!!) }
                .collect(Collectors.toSet()),
            friends.stream().map { f ->
                RelationDto(
                    f.id!!,
                    if (f.receiver?.id == user.id) f.sender!! else f.receiver!!,
                    f.createdTime?.toLocalDate()!!
                )
            }
                .collect(Collectors.toSet()),
            outgoingSubscriptions.stream().map { b -> RelationDto(b.id!!,b.receiver!!, b.createdTime?.toLocalDate()!!) }
                .collect(Collectors.toSet()),
            incomingSubscriptions.stream().map { b -> RelationDto(b.id!!,b.sender!!, b.createdTime?.toLocalDate()!!) }
                .collect(Collectors.toSet())
        )
    }
}